Automattic/newspack-plugin
# feat: segmentation criteria for subscriptions and memberships

### All Submissions:






### Changes proposed in this Pull Request:



Along with https://github.com/Automattic/newspack-popups/pull/1230, adds support and UI for two new sets of criteria, based on subscriptions and memberships.

Also splits the UI for the single segment screen into smaller subsections, as the current set of sections was looking crowded as we add more criteria. (cc @thomasguillot)

Note that I moved some files around and renamed some others for consistency. 

- The `ListsControl` component is only used in the context of the Campaigns wizard, so I moved it out of the Newspack Components directory and into the components for the wizard. Also made this more abstract so it could be reused for all of the `list__in` and `list__not_in` criteria.
- Renamed some of the components files in the Campaigns wizard folder from `CamelCase` to `kebab-case`, which matches most of our file naming across repos. Didn't try to be exhaustive about this.
- Made some minor style and markup updates to the `AutocompleteTokenfield` component, to make the placement of the spinner component during loading more consistent and predictable.

Closes `1205451924709229/1205603536115944`.

### How to test the changes in this Pull Request:

1. On `master`, register two new user accounts in separate sessions. On Account A, purchase a non-donation subscription product. On Account B, join any membership plan by satisfying the plan's requirements for granting access.
2. Check out this branch and https://github.com/Automattic/newspack-popups/pull/1230. Create two new segments at the highest priority, one for Subscriptions with "Has active subscription(s)" targeting the subscription product purchased by Account A, and one for Memberships with "Has active membership(s)" targeting the membership plan joined by Account B. Also note the new organization of the single segment screen here, which splits the "Reader Activity" section into three separate sections:

<img width="928" alt="Screenshot 2023-10-13 at 4 51 46 PM" src="https://github.com/Automattic/newspack-plugin/assets/2230142/de32dd57-cc18-4696-8e03-1b41cf7bcbd3">

3. Test support for segmenting user accounts who already have active subscriptions/memberships upon login. In separate sessions, log in as Account A and Account B and confirm that after logging in, each session gets segmented to the corresponding subscription/membership segment.
4. Test support for new subscriptions/memberships: In two new sessions, purchase the required subscription product and join the required membership plan. In each case confirm that you're segmented after completing the purchase/membership action.
5. Test support for changes subscription/membership statuses. As an admin, cancel the subscription/membership for the test users. Log in as those test users and confirm that you no longer match the segments.
6. Also test the new "Does not have active subscription/membership" criteria: the test users who have active subscriptions and memberships matching these criteria should not match these segments, while everyone else should.

### Other information:
